docs/source/notes/autograd.rst

```diff
@@ -1,124 +1,457 @@
+This is because a gradient of :math:`0` should propagate the non-finite values while an independent gradient should propagate an independent gradient (which is the same as :math:`0`).
+This is a problem when trying to hide pathological points with indexing as can be seen for example in `Issue 9688 <https://github.com/pytorch/pytorch/issues/9688>`_.
+
+How do we express gradients of zero
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A gradient of :math:`0` (or independent) can be expressed in any of the following ways:
+
+- A Tensor full of zeros
+- In Python, a ``None`` Tensor
+- In cpp, and ``undefined`` Tensor
+- Using ``autograd.grad(allow_unused=False)``, an error stating that the output is independent of the input
+
+Non-differentiable functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The chain rule can **only** be applied at points where all elementary operations
+are differentiable. If any of them is evaluated at a point where it is not,
+the computed gradient can be incorrect. For example, consider the gradient at :math:`0` of the identity function
+when it is decomposed as :math:`\text{id}(x) = \text{relu}(x) - \text{relu}(-x)`.
+
+This also means that the function that computes the :math:`2`-norm and the one that squares all the elements of a tensor, sums them and returns the square root
+of the result will not have the same behavior with respect to gradients all the time.
+In particular, if the sum is :math:`0`, the square root function is not differentiable at this point and so the returned gradient is undefined while the :math:`2`-norm function has a well defined gradient at :math:`0` of :math:`0`.
+
+To try and reduce the impact of this limitation, we define the gradients of the elementary operations by applying the following rules in order:
+
+#. If the function is not defined (:math:`\sqrt(-1)` or :math:`\log(-1)` for example) then this is undefined. Most function will return :math:`nan`, but for performance reasons, some functions will return not-nan values (:math:`\log(-1)` for example).
+#. If a gradient exist at the current point, return it.
+#. If the function is convex, return one of it's subgradients at the current point, if more than one exists, return the subgradient with minimum norm (as it is always a descent direction, see "Convex Optimization Algorithms" by Dimitri Bertsekas, Proposition 4.3.1 for a proof).
+#. Define the gradient at the current point by continuity (note that :math:`inf` is possible here). If multiple values are possible, pick one arbitrarily.
+
+.. note::
+
+    - Even though we try our best, unless all elementary operations used are in case 2, the returned gradient is not guaranteed to be correct.
+    - Both the ``detach()`` method and the ``torch.no_grad`` decorator are expected not to follow these rules.
+    - This is work in progress, if you find a :code:`Function` that does not follow these rules, please open a new issue with the tag "module:autograd".
+
+
+Implementation details
```

Comment:
this is new?  each subsection's title don't really tell me what does in each at a glance. i'd suggest using a question format again since it's a simple way of highlighting the message of each paragraph and also helps make each self-contained.